### ISSUE TITLE
Código fonte

### ISSUE THREAD

**Issue by ProjetoModelagem** [ProjetoModelagem/LAB_RNA#1]
Antes de rodar o código principal rode o código abaixo para subir as 4 arquivos testes do lab:

from google.colab import files
uploaded = files.upload()

depois disso só rodar o código abaixo:

import numpy as np
import matplotlib.pyplot as plt
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import MaxAbsScaler
from sklearn.metrics import mean_squared_error
import warnings
from sklearn.exceptions import ConvergenceWarning

warnings.filterwarnings("ignore", category=ConvergenceWarning)

file_paths = {
    "teste2": "teste2.npy",
    "teste3": "teste3.npy",
    "teste4": "teste4.npy",
    "teste5": "teste5.npy"
}

architectures = [
    (10, 5),
    (20, 10),
    (30, 15, 5)
]

executions = 10        # Número de execuções por arquitetura
max_iter = 1000        # Número máximo de iterações por treinamento

best_predictions = {}

for test_name, file_path in file_paths.items():
    arquivo = np.load(file_path, allow_pickle=True)
    x = arquivo[0]
    y_original = arquivo[1]

    # Normalização
    scaler = MaxAbsScaler().fit(y_original)
    y = np.ravel(scaler.transform(y_original))

    best_mse = float('inf')
    best_arch = None
    best_y_pred = None

    # Testa cada arquitetura
    for arch in architectures:
        for _ in range(executions):
            regr = MLPRegressor(
                hidden_layer_sizes=arch,
                max_iter=max_iter,
                activation='tanh',
                solver='adam',
                learning_rate='adaptive',
                early_stopping=True,
                n_iter_no_change=20,
                verbose=False
            )
            regr.fit(x, y)
            y_pred = regr.predict(x)
            mse = mean_squared_error(y, y_pred)

            if mse < best_mse:
                best_mse = mse
                best_arch = arch
                best_y_pred = y_pred

    # Salva o melhor resultado dessa base
    best_predictions[test_name] = {
        "x": x,
        "y_real": y,
        "y_pred": best_y_pred,
        "arch": best_arch,
        "mse": best_mse
    }

# Geração dos gráficos
fig, axs = plt.subplots(len(best_predictions), 3, figsize=(18, 4 * len(best_predictions)))
fig.suptitle("Melhores Resultados por Arquitetura e Base", fontsize=16)

for idx, (test_name, data) in enumerate(best_predictions.items()):
    x = data["x"]
    y_real = data["y_real"]
    y_pred = data["y_pred"]
    arch = data["arch"]
    mse = data["mse"]

    axs[idx, 0].plot(x, y_real)
    axs[idx, 0].set_title(f"{test_name} - Função Original")

    axs[idx, 1].plot(x, y_pred)
    axs[idx, 1].set_title(f"{test_name} - Predição\nArquitetura: {arch}")

    axs[idx, 2].plot(x, y_real, label="Original", linewidth=1)
    axs[idx, 2].plot(x, y_pred, label="Estimado", linewidth=2)
    axs[idx, 2].legend()
    axs[idx, 2].set_title(f"{test_name} - Comparação\nMSE: {mse:.6f}")

plt.tight_layout(rect=[0, 0, 1, 0.97])
plt.show()

resultados:

![Image](https://github.com/user-attachments/assets/7a060e65-61e1-4f7e-ab9b-42d3b39e5bae)
